server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java

```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
       );
     }
 
+    File tmpOutDir = new File(baseStorageDir, storageDir + "." + UUID.randomUUID().toString());
```

Comment:
I think a try-with-resources pattern could work here to ensure the temporary directory is cleaned up. Make an `AutoCloseable` (or `Closeable`) that deletes the temporary directory, and wrap the scope of the temporary directory in a `try(AutoCloseable foo = directoryDeletingCloseable)` or similar.